### PR TITLE
Spreed integration

### PR DESCRIPTION
**Pull request by leonklingele-work**

### CODE REVIEW

File: js/app/controllers/spreedmeetingcontroller.js [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+app.controller('SpreedMeetingController', ['$scope', '$http', '$timeout', function($scope, $http, $timeout) {
+	'use strict';
+
+	angular.extend($scope.properties, {
+		doScheduleMeeting: null, // TODO(leon): Retrieve from somewhere
+		roomToken: null, // TODO(leon): Retrieve from somewhere
+		oldRoomURL: null,
+	});
+
+	var attendeeRoles = {
```
Comment (tcitworld) [inline js/app/controllers/spreedmeetingcontroller.js]: Let's use `let`/`const` where it's possible. :)

File: js/app/controllers/editorcontroller.js [hunk 2]
```diff
@@ -53,15 +53,16 @@ app.controller('EditorController', ['$scope', 'TimezoneService', 'AutoCompletion
 			{title: t('calendar', 'Details'), value: 0},
 			{title: t('calendar', 'Attendees'), value: 1},
 			{title: t('calendar', 'Reminders'), value: 2},
-			{title: t('calendar', 'Repeating'), value: 3}
+			{title: t('calendar', 'Repeating'), value: 3},
+			{title: t('calendar', 'Spreed Meeting'), value: 4}
```
Comment (nickvergessen) [inline js/app/controllers/editorcontroller.js]: Spreed should not be used in the UI. It's "Video calls"

Comment (leonklingele-work) [inline js/app/controllers/editorcontroller.js]: So it should say 'Join video call'? Doesn't make sense to me. It could as well be a audio-only meeting.

Comment (nickvergessen) [inline js/app/controllers/editorcontroller.js]: Call Meeting?

At least not introduce a brand like name that is not used anywhere else in the user interface

Comment (bergart) [inline js/app/controllers/editorcontroller.js]: It is used under the Administration/Additional Settings: "Spreed video calls" for the STUN/TURN configuration

Comment (leonklingele-work) [inline js/app/controllers/editorcontroller.js]: Please report this here: https://github.com/nextcloud/spreed :)

File: templates/part.eventsattendees.php [hunk 3]
```diff
@@ -52,18 +52,16 @@
 					id="attendeecutype_{{$id}}">
 				</select>
 				<div class="attendeeopt pull-right">
-					<input
-						type="checkbox" class="attendeecheckbox event-checkbox"
+					<input type="checkbox" name="attendeecheckbox" class="checkbox"
```
Comment (leonklingele-work) [inline templates/part.eventsattendees.php]: All of the changes to this file should be made in a separate PR, not in this one.

File: templates/part.spreedmeeting.php [hunk 4]
```diff
@@ -28,11 +28,11 @@
 	// TODO(leon): Include link to app store (https://apps.nextcloud.com/apps/spreed)
 } else {?>
 
-<label>
-	<input
-		type="checkbox" class="spreedmeetingcreatecheckbox event-checkbox"
-		ng-model="properties.doScheduleMeeting" />
-	<?php p($l->t('Schedule Spreed Meeting for this event'));?>
+<input type="checkbox" name="spreedmeetingcreatecheckbox"
```
Comment (leonklingele-work) [inline templates/part.spreedmeeting.php]: Any reason the `label` no longer contains the `input` field? This requires us to introduce the `id` tag on `input`.

File: templates/part.spreedmeeting.php [hunk 5]
```diff
@@ -28,11 +28,11 @@
...
+	ng-model="properties.doScheduleMeeting"
+	id="spreedmeetingcreatecheckbox" class="checkbox" />
+<label for="spreedmeetingcreatecheckbox">
+	<?php p($l->t('Schedule Spreed Meeting for this event'))?>
```
Comment (leonklingele-work) [inline templates/part.spreedmeeting.php]: What happened to the trailing semicolon?

Comment (bergart) [reply templates/part.spreedmeeting.php]: you're right, fixed!

File: templates/part.spreedmeeting.php (hunk 4 above)
Comment (bergart) [reply templates/part.spreedmeeting.php]: most of the other code looks like that!

File: appinfo/application.php [hunk 6]
```diff
@@ -84,8 +84,10 @@ public function __construct($params=[]) {
 			$config = $c->getServer()->getConfig();
 			$urlGenerator = $c->getServer()->getURLGenerator();
 			$appManager = $c->getServer()->getAppManager();
+			$l10nFactory = $c->getServer()->getL10NFactory();
```
Comment (leonklingele-work) [inline appinfo/application.php]: I suggest this whole commit be made in a separate PR as it is unrelated to the _spreed-integration_, i.e. is also required for the Contacts app without Spreed.

Comment (leonklingele-work) [reply appinfo/application.php]: Optimally this change should land in the nextcloud/server directly and provide an API endpoint to retrieve all the languages. This code is basically a copy from the nextcloud settings app(?), we should be as DRY as possible.

**Comment by georgehrke** [comment]
Please fix the merge conflicts

File: tests/php/controller/contactcontrollerTest.php [hunk 7]
```diff
@@ -26,6 +26,8 @@ class ContactControllerTest extends \PHPUnit_Framework_TestCase {
 	private $appName;
 	private $request;
 	private $contacts;
+	private $config;
+	private $l10nFactory;
```
Comment (georgehrke) [inline tests/php/controller/contactcontrollerTest.php]: Please write actual tests here :)

File: templates/part.spreedmeeting.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
+<?php
+/**
+ * Calendar App
+ *
+ * @author Raghu Nayyar
+ * @author Georg Ehrke
+ * @copyright 2016 Raghu Nayyar <hey@raghunayyar.com>
+ * @copyright 2016 Georg Ehrke <oc.list@georgehrke.com>
```
Comment (georgehrke) [inline templates/part.spreedmeeting.php]: please add your own copyright ;)

**Comment by georgehrke** [comment]
There are quite a lot of Todos left ;)

File: controller/viewcontroller.php [hunk 9]
```diff
@@ -191,13 +199,20 @@ private function getTemplateParams() {
 		$defaultColor = $this->config->getAppValue('theming', 'color', '#0082C9');
 		$canSharePublicLink = $this->config->getAppValue('core', 'shareapi_allow_links', 'yes');
 
+		$isSpreedAvailable = $this->appManager->isEnabledForUser('spreed', null);
```
Comment (nickvergessen) [inline controller/viewcontroller.php]: should be done with events and be handled in the other app then

File: js/app/controllers/calcontroller.js [hunk 10]
```diff
@@ -295,6 +295,28 @@ app.controller('CalController', ['$scope', 'Calendar', 'CalendarService', 'VEven
 							});
 						}
 					}
+
+					// TODO(leon): Trigger event instead
```
Comment (nickvergessen) [inline js/app/controllers/calcontroller.js]: do the event?

File: js/app/controllers/calcontroller.js [hunk 11]
```diff
@@ -295,6 +295,28 @@ app.controller('CalController', ['$scope', 'Calendar', 'CalendarService', 'VEven
...
+					var spreedmeeting = event.getSimpleEvent().spreedmeeting;
+					if (spreedmeeting && spreedmeeting.parameters.token) {
+						var token = spreedmeeting.parameters.token;
+						var $icon = angular.element('<img>')
+							.attr('src', OC.filePath('spreed', 'img', 'app.svg'))
```
Comment (nickvergessen) [inline js/app/controllers/calcontroller.js]: `OC.imagePath()` is what you are looking for

File: js/app/controllers/editorcontroller.js (hunk 2 above)
Comment (nickvergessen) [reply js/app/controllers/editorcontroller.js]: This is still open, the app will be named "Talk" in 13, so it should refer to that, not spreed

File: js/app/controllers/spreedmeetingcontroller.js [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+app.controller('SpreedMeetingController', ['$scope', '$http', '$q', 'SpreedMeetingService', function($scope, $http, $q, SpreedMeetingService) {
```
Comment (nickvergessen) [inline js/app/controllers/spreedmeetingcontroller.js]: Move to spreed app?

File: js/app/controllers/calcontroller.js (hunk 10 above)
Comment (leonklingele-work) [reply js/app/controllers/calcontroller.js]: The Calendar app currently doesn't support to be extended from other apps, that's why the integration was made directly into this app.

File: controller/viewcontroller.php (hunk 9 above)
Comment (leonklingele-work) [reply controller/viewcontroller.php]: Same, currently not possible.

File: js/app/controllers/spreedmeetingcontroller.js (hunk 12 above)
Comment (leonklingele-work) [reply js/app/controllers/spreedmeetingcontroller.js]: Same, not possible at the moment.

File: controller/viewcontroller.php (hunk 9 above)
Comment (nickvergessen) [reply controller/viewcontroller.php]:
```php
$this->dispatcher->dispatch('<event name>', new GenericEvent($data));
```

anyway…

Comment (leonklingele-work) [reply controller/viewcontroller.php]: > currently not possible

What I meant by that is that the Calendar app can't be extended by the Spreed app. Triggering an event wouldn't help at all since there's nothing to be done by Spreed.

Comment (nickvergessen) [reply controller/viewcontroller.php]: Spreed could extend the array before it's send to the Template, anyway

Comment (georgehrke) [reply controller/viewcontroller.php]: That's indeed true for loading the template variables. But right now there is no way for the spreed app to register a new tab in the event editor.

**Comment by georgehrke** [comment]
@leonklingele-work Can you please add before / after screenshots for GUI elements that changed and / or were added? 

cc @jancborchardt

File: js/app/controllers/attendeecontroller.js [hunk 13]
```diff
@@ -114,7 +115,8 @@ app.controller('AttendeeController', function($scope, AutoCompletionService) {
 				role: 'REQ-PARTICIPANT',
 				rsvp: 'TRUE',
 				partstat: 'NEEDS-ACTION',
-				cutype: 'INDIVIDUAL'
+				cutype: 'INDIVIDUAL',
+				isregistered: item.is_local_user ? 'TRUE' : 'FALSE'
```
Comment (georgehrke) [inline js/app/controllers/attendeecontroller.js]: You can use `true` and `false` here. 
RSVP is a string, because that item is part of the calendar data and encoded as a string there

File: js/app/controllers/calcontroller.js [hunk 14]
```diff
@@ -26,8 +26,8 @@
 * Description: The fullcalendar controller.
 */
 
-app.controller('CalController', ['$scope', 'Calendar', 'CalendarService', 'VEventService', 'SettingsService', 'TimezoneService', 'VEvent', 'is', 'fc', 'EventsEditorDialogService', 'PopoverPositioningUtility', '$window', 'isPublic', 'constants',
-	function ($scope, Calendar, CalendarService, VEventService, SettingsService, TimezoneService, VEvent, is, fc, EventsEditorDialogService, PopoverPositioningUtility, $window, isPublic, constants) {
+app.controller('CalController', ['$scope', '$location', 'Calendar', 'CalendarService', 'VEventService', 'SettingsService', 'TimezoneService', 'VEvent', 'is', 'fc', 'EventsEditorDialogService', 'PopoverPositioningUtility', '$window', 'isPublic', 'constants', 'SpreedMeetingService',
+	function ($scope, $location, Calendar, CalendarService, VEventService, SettingsService, TimezoneService, VEvent, is, fc, EventsEditorDialogService, PopoverPositioningUtility, $window, isPublic, constants, SpreedMeetingService) {
```
Comment (georgehrke) [inline js/app/controllers/calcontroller.js]: Where do you actually use `$location` here?

File: js/app/models/simpleEventModel.js [hunk 15]
```diff
@@ -41,7 +41,8 @@ app.factory('SimpleEvent', function () {
 		'repeating': null,
 		'rdate': null,
 		'rrule': null,
-		'exdate': null
+		'exdate': null,
+		'spreedmeeting': null
```
Comment (georgehrke) [inline js/app/models/simpleEventModel.js]: `spreedmeeting` is not part of RFC5545 ;)
Please use `X-SPREED-MEETING` as noted earlier already

File: js/app/models/simpleEventModel.js [hunk 16]
```diff
@@ -51,13 +52,20 @@ app.factory('SimpleEvent', function () {
 		'cutype',
 		'cn',
 		'delegated-from',
-		'delegated-to'
+		'delegated-to',
+		'spreedmeetingrole',
```
Comment (georgehrke) [inline js/app/models/simpleEventModel.js]: `X-`...

File: js/app/models/simpleEventModel.js [hunk 17]
```diff
@@ -51,13 +52,20 @@ app.factory('SimpleEvent', function () {
...
+		'isregistered'
```
Comment (georgehrke) [inline js/app/models/simpleEventModel.js]: `X-`...

File: js/app/service/spreedmeetingservice.js [hunk 18]
```diff
@@ -0,0 +1,111 @@
+/**
+ * Calendar App
+ *
+ * @author Raghu Nayyar
+ * @author Georg Ehrke
+ * @copyright 2016 Raghu Nayyar <hey@raghunayyar.com>
+ * @copyright 2016 Georg Ehrke <oc.list@georgehrke.com>
```
Comment (georgehrke) [inline js/app/service/spreedmeetingservice.js]: your own copyright

File: templates/part.spreedmeeting.php [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+?>
+
+<?php if (!$_['isSpreedAvailable']) {
+	p($l->t('The Spreed "Video calls" app is not available. Please ask your system administrator to install it.'));
+	// TODO(leon): Include link to app store (https://apps.nextcloud.com/apps/spreed)
```
Comment (georgehrke) [inline templates/part.spreedmeeting.php]: This tab should never be shown if spreed is not installed anyway

**Comment by jancborchardt** [comment]
(Yes please - all pull requests with visual changes should have before/after screenshots so changes can be confirmed to be as expected. :)

**Comment by ChristophWurst** [comment]
Ouch, that one didn't survive the Vue rewrite judging by the diff view here. I'll close this one as the chances of merging are really low in this current state.
